### PR TITLE
Synchronize search parameter changes between instances

### PR DESCRIPTION
**Pull request by feordin**
## Description
Use the ReindexJobWorker to poll for SearchParameter changes made by other instances

## Related issues
Addresses [issue [AB#81849](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/81849)].

## Testing
Manual testing by running multiple instances locally

## FHIR Team Checklist
- [X] **Update the title** of the PR to be succinct and less than 50 characters
- [X] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [X] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [X] Tag the PR with **Azure API for FHIR** if this will release to the managed service
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)
Skip

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/SearchParameterOperations.cs [hunk 1]
```diff
@@ -176,5 +188,86 @@ public async Task UpdateSearchParameterAsync(ITypedElement searchParam, RawResou
...
+            var paramsToAdd = new List<ITypedElement>();
+            foreach (var searchParam in updatedSearchParameterStatus.Where(p => p.Status != SearchParameterStatus.Deleted))
+            {
+                var searchParamResource = await GetSearchParameterByUrl(searchParam.Uri.ToString(), cancellationToken);
+
+                if (searchParamResource == null)
+                {
+                    _logger.LogError(
+                        "Updated SearchParameter status found for SearchParameter: {0}, but did not find any SearchParameter resources when querying for this url.",
+                        searchParam.Uri);
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/SearchParameterOperations.cs:220]: Perhaps a warning instead?

```suggestion
                    _logger.LogWarning(
                        "Updated SearchParameter status found for SearchParameter: {0}, but did not find any SearchParameter resources when querying for this url.",
                        searchParam.Uri);
```

File: src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/SearchParameterOperations.cs [hunk 2]
```diff
@@ -176,5 +188,86 @@ public async Task UpdateSearchParameterAsync(ITypedElement searchParam, RawResou
...
+            }
+        }
+
+        private void DeleteSearchParameter(string url)
+        {
+            try
+            {
+                _searchParameterDefinitionManager.DeleteSearchParameter(url, false);
+            }
+            catch (Exception ex) when (ex.Message.Contains("was not found", StringComparison.OrdinalIgnoreCase))
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/SearchParameterOperations.cs:250]: Can the error type be more specific?

File: src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/SearchParameterOperations.cs [hunk 3]
```diff
@@ -176,5 +188,86 @@ public async Task UpdateSearchParameterAsync(ITypedElement searchParam, RawResou
...
+            using IScoped<ISearchService> search = _searchServiceFactory.Invoke();
+            var queryParams = new List<Tuple<string, string>>();
+
+            queryParams.Add(new Tuple<string, string>("url", url));
+            var result = await search.Value.SearchAsync(KnownResourceTypes.SearchParameter, queryParams, cancellationToken);
+
+            if (result.Results.Any())
+            {
+                // There should be only one SearchParameter per url
+                return result.Results.First().Resource.RawResource.ToITypedElement(_modelInfoProvider);
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/SearchParameterOperations.cs:267]: Should we assert or log something if more than one? What is the impact?

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchOptionsFactory.cs [hunk 4]
```diff
@@ -333,7 +333,8 @@ public SearchOptions Create(string compartmentType, string compartmentId, string
             if (unsupportedSearchParameters.Any())
             {
                 bool throwForUnsupported = false;
-                if (_contextAccessor.RequestContext.RequestHeaders != null &&
+                if (_contextAccessor.RequestContext != null &&
+                    _contextAccessor.RequestContext.RequestHeaders != null &&
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchOptionsFactory.cs:337]: Can we combine:
```suggestion
                if (_contextAccessor?.RequestContext.RequestHeaders != null &&
```

**Review by brendankowitz — COMMENTED** [review]
Are these updates also reflected in the /metadata endpoint?

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Reindex/ReindexJobWorker.cs [hunk 5]
```diff
@@ -45,45 +57,71 @@ public async Task ExecuteAsync(CancellationToken cancellationToken)
 
             while (!cancellationToken.IsCancellationRequested)
             {
-                try
+                if (_searchParametersInitialized)
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Operations/Reindex/ReindexJobWorker.cs:60]: Should there be a delay in this loop while it waits for this?

**Review by brendankowitz — APPROVED** [review]

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Reindex/ReindexJobWorker.cs (hunk 5 above)
Comment (feordin) [reply src/Microsoft.Health.Fhir.Core/Features/Operations/Reindex/ReindexJobWorker.cs:60]: Yes! That is a good idea.

**Pull request merged**